modules/objdetect/src/qrcode.cpp

```diff
@@ -2367,22 +2436,49 @@ bool QRDecode::versionDefinition()
-                                     (version == 3) ? 1.5 :
-                                     version * (version + 1);
-    const Size newFactorSize(
-                  cvRound(no_border_intermediate.size().width  * multiplyingFactor),
-                  cvRound(no_border_intermediate.size().height * multiplyingFactor));
+    double multiplyingFactor = (version < 3)  ? 1 :
+                               (version == 3) ? 1.5 :
+                                version * 5.;
+    if (original_curved_points.empty() && version > 2)
+    {
+        multiplyingFactor = 1.0;
+        test_perspective_size = 4*version_size+1;
+        const Size temporary_size(cvRound(test_perspective_size), cvRound(test_perspective_size));
+        vector<Point2f> perspective_points;
+        perspective_points.push_back(Point2f(0.f, 0.f));
+        perspective_points.push_back(Point2f(test_perspective_size, 0.f));
+        perspective_points.push_back(Point2f(test_perspective_size, test_perspective_size));
+        perspective_points.push_back(Point2f(0.f, test_perspective_size));
+        perspective_points.push_back(Point2f(test_perspective_size * 0.5f, test_perspective_size * 0.5f));
+        vector<Point2f> pts = original_points;
+        const Point2f centerPt = intersectionLines(original_points[0], original_points[2],
+                                                   original_points[1], original_points[3]);
+        pts.push_back(centerPt);
+        Mat H = findHomography(pts, perspective_points);
+        Mat temp_intermediate; // Todo: check bin_barcode, this Mat not bin
+        warpPerspective(bin_barcode, temp_intermediate, H, temporary_size, INTER_NEAREST);
+        no_border_intermediate = temp_intermediate(Range(1, temp_intermediate.rows), Range(1, temp_intermediate.cols));
+        //imwrite("no_border_intermediate_size.png", no_border_intermediate);
+    }
+    int w = cvRound(no_border_intermediate.size().width * multiplyingFactor);
+    int h = cvRound(no_border_intermediate.size().width * multiplyingFactor);
+    if (version >= 2) { // Todo: fix tests
+        w += w % version_size;
+        h += h % version_size;
+    }
+    const Size newFactorSize(w, h);
     Mat postIntermediate(newFactorSize, CV_8UC1);
     resize(no_border_intermediate, postIntermediate, newFactorSize, 0, 0, INTER_AREA);
+    // cv::imwrite("postIntermediate.png", postIntermediate);
+    // cv::imwrite("tmp.png", tmp);
 
     const int delta_rows = cvRound((postIntermediate.rows * 1.0) / version_size);
     const int delta_cols = cvRound((postIntermediate.cols * 1.0) / version_size);
 
     vector<double> listFrequencyElem;
-    for (int r = 0; r < postIntermediate.rows; r += delta_rows)
+    for (int i = 0, r = 0; i < version_size; i++ && r < postIntermediate.rows, r+= delta_rows)
     {
-        for (int c = 0; c < postIntermediate.cols; c += delta_cols)
+        for (int j = 0, c = 0; j < version_size; j++ && c < postIntermediate.cols, c+= delta_cols)
```

Comment:
Overcomplicated (need to rewrite) or just invalid "for" loops.

Reply:
fixed